mlx/ops.cpp

```diff
@@ -2758,4 +2760,91 @@ array dequantize(
   return w_full;
 }
 
+std::map<char, int> string_to_char_map(std::string inp) {
+  std::map<char, int> counts;
+  for (int i = 0; i < inp.size(); i++) {
+    counts[inp[i]] = i;
+  }
+  return counts;
+}
+
+array einsum(
+    std::string equation,
+    const std::vector<array>& operands,
+    StreamOrDevice s /* = {} */) {
+  std::string output;
+  std::string input = equation;
+  if (equation.find("->") == std::string::npos) {
+    output = std::string(equation);
+    std::sort(output.begin(), output.end());
+  } else {
+    auto pos = equation.find("->");
+    output = equation.substr(pos + 2);
+    input = equation.substr(0, pos);
+  }
+  std::vector<std::string> inputs;
+  std::stringstream ss(input);
+  std::string token;
+  while (getline(ss, token, ',')) {
+    inputs.push_back(token);
+  }
+  if (operands.size() != inputs.size()) {
+    throw std::runtime_error(
+        "Number of operands must match the number of input characters");
+  }
+
+  std::map<char, int> input_map;
+  for (int i = 0; i < inputs.size(); i++) {
+    auto arr = operands[i];
+    auto inp = inputs[i];
+    for (int j = 0; j < MIN(arr.shape().size(), inp.size()); j++) {
```

Comment:
I think std:min is better than a #define for this (at least consistent with our style)